server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java

```diff
@@ -544,21 +556,97 @@ public String toString() {
 
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        Metadata.XContentContext context = Metadata.XContentContext.valueOf(params.param(CONTEXT_MODE_PARAM, CONTEXT_MODE_API));
         builder.startObject(getId());
-        builder.field("name", getName());
-        builder.field("ephemeral_id", getEphemeralId());
-        builder.field("transport_address", getAddress().toString());
+        builder.field(KEY_NAME, getName());
+        builder.field(KEY_EPHEMERAL_ID, getEphemeralId());
+        builder.field(KEY_TRANSPORT_ADDRESS, getAddress().toString());
 
-        builder.startObject("attributes");
+        builder.startObject(KEY_ATTRIBUTES);
         for (Map.Entry<String, String> entry : attributes.entrySet()) {
             builder.field(entry.getKey(), entry.getValue());
         }
         builder.endObject();
+        if (context == Metadata.XContentContext.GATEWAY) {
+            builder.field(KEY_HOST_NAME, getHostName());
+            builder.field(KEY_HOST_ADDRESS, getHostAddress());
+            builder.field(KEY_VERSION, getVersion().toString());
+            builder.startArray(KEY_ROLES);
+            for (DiscoveryNodeRole role : roles) {
+                builder.value(role.roleName());
+            }
+            builder.endArray();
+        }
 
         builder.endObject();
         return builder;
     }
 
+    public static DiscoveryNode fromXContent(XContentParser parser, String nodeId) throws IOException {
+        if (parser.currentToken() == null) {
+            parser.nextToken();
+        }
+        if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+            parser.nextToken();
+        }
+        if (parser.currentToken() != XContentParser.Token.FIELD_NAME) {
+            throw new IllegalArgumentException("expected field name but got a " + parser.currentToken());
+        }
+        String nodeName = null;
+        String hostName = null;
+        String hostAddress = null;
+        String ephemeralId = null;
+        TransportAddress transportAddress = null;
+        Map<String, String> attributes = new HashMap<>();
+        Set<DiscoveryNodeRole> roles = new HashSet<>();
+        Version version = null;
+        String currentFieldName = parser.currentName();
+        // token should be start object at this point
+        // XContentParser.Token token = parser.nextToken();
+        // if (token != XContentParser.Token.START_OBJECT) {
+        // throw new IllegalArgumentException("expected object but got a " + token);
+        // }
+        XContentParser.Token token;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
```

Comment:
This parsing isn't very readable. You have multi-branch if/else wrapped inside other multi-branch if/else.  I'm sure somewhere there's a complexity metric that would scream at you. :)

You always will have a field name (or it's invalid XContent) so if you're expecting the field name you should just be able to get the name. 

After that you can use a switch statement on the field names since you know their data types as well (or it's invalid and will throw an exception if it's the wrong type).

Something like
```java
while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
    String fieldName = parser.currentName();
    parser.nextToken();
    switch (fieldName) {
        case KEY_NAME:
            nodeName = parser.text();
            break;
        case KEY_EPHEMERAL_ID:
            ephemeralId = parser.text();
            break;
        // and so on.
```

Same comment on `DiscoveryNodes` class below.

Reply:
I agree @dbwiddis, I was bothered by these nested if/else, but that's how we were doing this in many other classes, so i didn't change the approach. But, will update this to use switch statement like you suggested.